### PR TITLE
timestamp gitlab ci job outputs

### PR DESCRIPTION
**Pull request by altaua**
cf https://github.com/paritytech/ci_cd/issues/715
based on previous work by @alvicsam in #13047

### CODE REVIEW

**Comment by oleg-plakida** [comment]
I don't like an optional timestamp functionality. This adds complexity. I'd rather find the way to fix the issues with some pipelines failing with this script.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2521842

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2522304

**Comment by altaua** [comment]
Turns out the linux-stable-int test was just a hardcoded timeout inside the test being hit because the naive timestamping script we'd been using slows output-heavy tests down a lot.

This was further compounded by the old script being bugged and discarding all remaining buffered output when the job script finishes.

File: scripts/ci/gitlab/timestamp.yml [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+        cleanup() {
+          exec >/dev/null 2>&1
+          wait "${TIMESTAMP_PID}"
+        }
+
+        exec > >(s_datestamp) 2>&1
+
+        TIMESTAMP_PID=$!
+        trap cleanup EXIT
```
Comment (oleg-plakida) [inline scripts/ci/gitlab/timestamp.yml:26]: Interesting. I don't understand what is it for) Could you please explain?

**Review by oleg-plakida — APPROVED** [review]

File: scripts/ci/gitlab/timestamp.yml (hunk 1 above)
Comment (altaua) [reply scripts/ci/gitlab/timestamp.yml:26]: Oh, that's to fix a bug in the original script: The main job script might finish before the s_datestamp process substitution has had a chance to process all output; if we don't wait for it, bash will just exit immediately, killing s_datestamp and thus discarding the last chunk of output.

Comment (altaua) [reply scripts/ci/gitlab/timestamp.yml:26]: (Incidentally, that's what made us not see the timeout error message in test-linux-stable-int for #13047)

**Review by alvicsam — APPROVED** [review]

**Comment by oleg-plakida** [comment]
It might be useful to add this script to other projects. Do you know is there any shared library for scripts in the gitlab?

**Review by rcny — APPROVED** [review]

**Comment by rcny** [comment]
> > It might be useful to add this script to other projects. Do you know is there any shared library for scripts in the gitlab?
> 
> Other repos could include the .timestamp spec with an [include:project](https://docs.gitlab.com/ee/ci/yaml/#includeproject) directive; we just need to discuss if we're fine with including it from substrate or would like to start a dedicated pipeline library repo oslt (or do we perchance have one and I just haven't discovered it yet?).

I've created [this project](https://gitlab.parity.io/parity/infrastructure/ci_cd/shared). Feel free to add the `timestamp` snippet there and reference it here.

**Comment by oleg-plakida** [comment]
> > It might be useful to add this script to other projects. Do you know is there any shared library for scripts in the gitlab?
> 
> Other repos could include the .timestamp spec with an [include:project](https://docs.gitlab.com/ee/ci/yaml/#includeproject) directive; we just need to discuss if we're fine with including it from substrate or would like to start a dedicated pipeline library repo oslt (or do we perchance have one and I just haven't discovered it yet?).

The dedicated library sounds just right. This would help us to maintain shared code properly and use DRY approach.

**Comment by altaua** [comment]
.timestamp will have to remain disabled for macos jobs; `/bin/bash` on ci-mac2 is far too old, and annoyingly gitlab can't be configured to use the homebrew bash version instead.

For reference, the missing bash features are
- the `printf %(..)T` format specifier -- this could be worked around by invoking `date` instead on macos, that'd just add some overhead (observed up to 100% for output-heavy jobs)
- writing process substitution pids to $? -- working around this would be very annoying, and if we don't, the jobs would be at risk of not printing parts of their output.

**Comment by altaua** [comment]
bot merge

**Pull request merged**